[utils] Move write_json to write_transactionally

jimporter opened the pull request:
99% of write_json's implementation is about file I/O, so this makes more sense in FileOps. It will also make it easier for Boost.JSON to use this function eventually.

ricab reviewed (commented):
Makes sense to me @jimporter. It clarifies the code a bunch, separates concerns better, and makes it more reusable. Only a small detail and an optional solicitation, if you're in feeling Christmas approaching 🎅 

[secondary]

jimporter commented on tests/test_base_snapshot.cpp:
Good catch! (Though note that this will get simplified even further in an [upcoming commit](https://github.com/canonical/multipass/commit/5b5917bfcde589fc59bc1c59e4b2d66cc17566d1#diff-1dda53f0bcd7bfaeedddf0c9d5b4aadad0fbdafb303ee8d59dca420b88aaf9eaR393).)

jimporter commented on src/utils/file_ops.cpp:
```diff
@@ -38,6 +50,89 @@ mp::FileOps::FileOps(const Singleton<FileOps>::PrivatePass& pass) noexcept
 {
 }
 
+void mp::FileOps::write_transactionally(const QString& file_name, const QByteArray& data) const
```
Ok, done. I pulled the backoff-timing logic into a separate class, which simplifies that code a bit (and keeps us from repeatedly instantiating new `std::random_device`s every call). I also fixed a small bug in the implementation where we'd call `sleep_for` after all our attempts were exhausted. We can just throw an exception immediately in that case.

ricab commented on src/utils/file_ops.cpp:
```diff
@@ -16,13 +16,55 @@
 
+namespace
+{
+constexpr static auto kLogCategory = "FileOps";
```
Forgot old name?

ricab commented on src/utils/file_ops.cpp:
```diff
@@ -16,13 +16,55 @@
+    }
+
+    // Delay with jitter + backoff. A typical series produced by this formula would look like as
+    // follows:
```
```suggestion
    // Delay with jitter + backoff. Typical series produced by this formula would be:
```

jimporter replied:
Done.

ricab commented on src/utils/file_ops.cpp:
```diff
@@ -38,6 +80,79 @@ mp::FileOps::FileOps(const Singleton<FileOps>::PrivatePass& pass) noexcept
+        std::this_thread::sleep_for(delay);
+    }
+
+    assert(false && "We should never get here");
```
👍

ricab reviewed (commented):
Yep, well done. Only a few remaining details.

jimporter commented on src/utils/file_ops.cpp:
```diff
@@ -16,13 +16,55 @@
+class BackoffTimer
+{
+public:
+    using duration = std::chrono::milliseconds;
```
Fixed. (I'm sure I'll forget about that again at some point since I'm *very* used to camel case being used for template identifiers.)

ricab reviewed (dismissed):
Yep, I am happy with this if CI is.

jimporter commented:
@ricab I updated this a bit, and added another commit at the beginning. That one makes all our `FileOps` wrappers for `QFile` use the original type for the file argument, which makes it easier to use those functions as needed (specifically, this helps because I changed `write_transactionally` to take a `QByteArrayView` instead of a `QByteArray`, and so I used a different overload of `FileOps::write`).

tobe2098 commented on src/utils/file_ops.cpp:
```diff
@@ -16,14 +16,52 @@
+public:
+    using Duration = std::chrono::milliseconds;
+    BackoffTimer(Duration factor, Duration max_duration, Duration max_jitter)
+        : rng(std::random_device{}()),
```
Is there a reason we need a truly random number generator when available in hardware? Each generation is a syscall in those devices. Would a pseudo rng like std::mt19937_64 be sufficient for this?

jimporter replied:
This is just to set the seed for our PRNG. How about making the PRNG `thread_local` so we don't have to recreate it as often?
